File: Homework5/problem2.jl
using LinearAlgebra
using Plots
using DelimitedFiles

function RK4(f::Function, t_start::Float64, t_end::Float64, y_init, steps::Int)
    y = zeros((length(y_init), steps))
    h = (t_end - t_start) / steps
    t = t_start
    y[:, 1] = y_init
    for i in 1:steps-1
        k1 = f(t, y[:, i])
        k2 = f(t + 0.5 * h, y[:, i] .+ 0.5 .* h .* k1)
        k3 = f(t + 0.5 * h, y[:, i] .+ 0.5 .* h .* k2)
        k4 = f(t + h, y[:, i] .+ h .* k3)
        y[:, i + 1] = y[:, i] .+ h ./ 6. .* (k1 + 2 .* k2 + 2 .* k3 + k4) 
        t += h
    end
    return y
end

function astronomy(t, y)
    mu = 0.012277471
    mu_h = 1 - mu
    result = zeros((length(y),))
    D1 = ((y[1] + mu) ^ 2 + y[3] ^ 2) ^ 1.5
    D2 = ((y[1] - mu_h) ^ 2 + y[3] ^ 2) ^ 1.5
    result[1] = y[2]
    result[2] = y[1] + 2 * y[4] - mu_h * (y[1] + mu) / D1 - mu * (y[1] - mu_h) / D2 
    result[3] = y[4]
    result[4] = y[3] - 2 * y[2] - mu_h * y[3] / D1 - mu * y[3] / D2
    return result
end

y_init = zeros((4,))
y_init[1] = 0.994
y_init[2] = 0.0
y_init[3] = 0.0
y_init[4] = -2.00158510637908252240537862224
t_start = 0.0
t_end = 17.1
steps = 
for steps in [1000, 5000, 10000, 50000, 100000]
    y = RK4(astronomy, t_start, t_end, y_init, steps)
    display(plot(y[1,:], y[3,:],
                xlabel="u_1",
                ylabel="u_2",
                title=string("steps = ", steps)))
    png(string("problem2_", steps, ".png"))
end

File: Homework5/problem2.m
close all

y_init = [0.994; 0.0; 0.0; -2.00158510637908252240537862224];
tspan = [0.0, 17.1];
[t, y] = ode45(@threebody, tspan, y_init);
figure
plot(y(:,1), y(:, 3))
xlabel('u_1')
ylabel('u_2')
title('Solution using ode45')


steps = 1000;
h = (tspan(2) - tspan(1)) / steps;
[t_rk, y_rk] = rk4(@threebody, tspan, y_init, h);
figure
plot(y_rk(1,:), y_rk(3,:))

function [t,y] = rk4(f,tspan,y0,h)
y0 = y0(:); % make sure y0 is a column vector
m = length(y0); % problem size
t = tspan(1):h:tspan(2); % output abscissae
N = length(t)-1; % number of steps
y = zeros(m,N+1);
y(:,1) = y0; % initialize
% Integrate
for i=1:N
% Calculate the four stages
K1 = feval(f, t(i),y(:,i) );
K2 = feval(f, t(i)+.5*h, y(:,i)+.5*h*K1);
K3 = feval(f, t(i)+.5*h, y(:,i)+.5*h*K2);
K4 = feval(f, t(i)+h, y(:,i)+h*K3 );
% Evaluate approximate solution at next step
y(:,i+1) = y(:,i) + h/6 *(K1+2*K2+2*K3+K4);
end
end

function [ dydt ] = threebody( t, y )
%THREEBODY Summary of this function goes here
%   Detailed explanation goes here
    mu = 0.012277471;
    mu_h = 1 - mu;
    D1 = ((y(1) + mu) ^ 2 + y(3) ^ 2) ^ 1.5;
    D2 = ((y(1) - mu_h) ^ 2 + y(3) ^ 2) ^ 1.5;
    dydt(1) = y(2);
    dydt(2) = y(1) + 2 * y(4) - mu_h * (y(1) + mu) / D1 - mu * (y(1) - mu_h) / D2 ;
    dydt(3) = y(4);
    dydt(4) = y(3) - 2 * y(2) - mu_h * y(3) / D1 - mu * y(3) / D2;
    dydt = dydt';
end